pkg/koordlet/prediction/peak_predictor.go

```diff
@@ -0,0 +1,168 @@
+	ProdReclaimablePredictor PredictorType = iota
+)
+
+// PredictorFactory is an interface for creating predictors of different types.
+type PredictorFactory interface {
+	New(PredictorType) Predictor
+}
+
+type Predictor interface {
+	AddPod(pod *v1.Pod) error
+
+	GetResult() (v1.ResourceList, error)
+}
+
+type predictorFactory struct {
+	predictServer     PredictServer
+	coldStartDuration time.Duration
+}
+
+// NewPredictorFactory creates a new instance of PredictorFactory.
+func NewPredictorFactory(predictServer PredictServer, coldStartDuration time.Duration) PredictorFactory {
+	return &predictorFactory{
+		predictServer:     predictServer,
+		coldStartDuration: coldStartDuration,
+	}
+}
+
+// New creates a new instance of a predictor based on the given type.
+func (f *predictorFactory) New(t PredictorType) Predictor {
+	switch t {
+	case ProdReclaimablePredictor:
+		return &prodReclaimablePredictor{
+			predictServer:     f.predictServer,
+			coldStartDuration: f.coldStartDuration,
+			reclaimable:       util.NewZeroResourceList(),
+			pods:              make(map[string]bool),
+		}
+	default:
+		return &emptyPredictor{}
+	}
+}
+
+type emptyPredictor struct {
+}
+
+// AddPod adds a pod to the predictor for resource prediction.
+func (p *emptyPredictor) AddPod(pod *v1.Pod) error {
+	return nil
+}
+
+// GetResult returns an error indicating that the predictor is empty.
+func (p *emptyPredictor) GetResult() (v1.ResourceList, error) {
+	return nil, fmt.Errorf("empty pridictor")
+}
+
+func NewEmptyPredictorFactory() PredictorFactory {
+	return &emptyPredictorFactory{}
+}
+
+type emptyPredictorFactory struct {
+}
+
+func (f *emptyPredictorFactory) New(t PredictorType) Predictor {
+	return &emptyPredictor{}
+}
+
+type prodReclaimablePredictor struct {
+	predictServer     PredictServer
+	coldStartDuration time.Duration
+
+	reclaimable v1.ResourceList
+	pods        map[string]bool
+}
+
+// AddPod adds a pod to the predictor for resource prediction.
+func (p *prodReclaimablePredictor) AddPod(pod *v1.Pod) error {
+	if p.pods[string(pod.UID)] {
+		return fmt.Errorf("Pod %v already exists", pod.UID)
+	} else {
+		p.pods[string(pod.UID)] = true
+	}
+
+	// Pods in cold start have 0 reclaimable resources
+	if time.Since(pod.CreationTimestamp.Time) <= p.coldStartDuration {
+		return nil
+	}
+
+	// Pods in terminating stage have 0 reclaimable resources
+	if pod.DeletionTimestamp != nil {
+		return nil
+	}
+
+	result, err := p.predictServer.GetPrediction(MetricDesc{UID: UIDType(pod.UID)})
+	if err != nil {
+		return err
+	}
+	p95Resources := result.Data["p95"]
+	p98Resources := result.Data["p98"]
+
+	podRequests := util.GetPodRequest(pod, v1.ResourceCPU, v1.ResourceMemory)
+	podCPURequest := podRequests[v1.ResourceCPU]
+	podMemoryRequest := podRequests[v1.ResourceMemory]
+
+	reclaimableCPUMilli := int64(0)
+	reclaimableMemoryBytes := int64(0)
+
+	if p95CPU, ok := p95Resources[v1.ResourceCPU]; ok {
+		peakCPU := util.MultiplyMilliQuant(p95CPU, 1.1)
```

Comment:
This ratio can be inaccurate and not universal. How about adding a parameter to adjust it when necessary?

Reply:
later add a pr separately to add configurations.